### ISSUE TITLE
AOF rewrite not working

### ISSUE THREAD

**Issue by ppanyukov** [microsoftarchive/redis#275]
This is probably related to https://github.com/MSOpenTech/redis/issues/271 but I'm logging this as a separate issue.

It surfaced from the SO question: http://stackoverflow.com/questions/31525903/redis-aborting-for-out-of-memory, the same one that prompted https://github.com/MSOpenTech/redis/issues/274.

Following the advice to use large `maxheap` the OOM issue is gone but now we are getting AOF rewrite not working.

The command I'm starting redis with is:

```
redis-server --loglevel verbose --maxheap 500000000 --maxmemory 134217728 --maxmemory-policy allkeys-lru --appendonly yes --hz 10 --save '""' --auto-aof-rewrite-percentage 100 --auto-aof-rewrite-min-size 64mb --no-appendfsync-on-rewrite yes
```

What I'm seeing in the logs is:

```
[19284] 24 Jul 16:18:59.247 # fork operation complete
[19284] 24 Jul 16:18:59.252 # Background AOF rewrite terminated with error
[19284] 24 Jul 16:18:59.352 * Starting automatic rewriting of AOF on 10010869700% growth
[19284] 24 Jul 16:18:59.356 * Background append only file rewriting started by pid 30536
[30536] 24 Jul 16:18:59.399 # Write error writing append only file on disk: Invalid argument
[30536] 24 Jul 16:18:59.400 # rewriteAppendOnlyFile failed in qfork: Invalid argument
```

Interestingly, the issue is not there if I compile in debug mode -- surely a sign of memory corruption somewhere?

After debugging the child process (started as `redis-server.exe --qfork 252 27516 --logfile "stdout"`) I got it go boom here:

```
// aof.c, line 1010
        d = db->dict;
```

There looks to be an issue with `db->dict` pointer.

Here is what my VS debugger window looks like:

![image](https://cloud.githubusercontent.com/assets/300031/8878006/867121d0-3220-11e5-8f4c-c15a2fea23b2.png)

Several outcomes at this point:
1. We hit `VectoredHeapMapper` handler which resolves the access violation and we continue.
2. Same as above but then we get BOOM and the following 

```
First-chance exception at 0x0000000140033B2A in redis-server.exe: 0xC0000005: Access violation reading location 0x00007FF5ED218CE0.
```

I should say that we may get further than that line `d = db->dict;` and hit the handler many times before the crash.

Where and how exactly the crash is happening I'm not sure.

Like I said, things work better when compiled in Debug mode, given that you can avoid OOM.

The very nasty side effect of AOF rewrite not working is that the aof file keeps growing and growing and growing. Until you run out of disk space I guess?

**Comment by enricogior** [comment]
Hi @ppanyukov,
a quick answer:
- We hit VectoredHeapMapper handler which resolves the access violation and we continue.

that is perfecctly normal, it's just virtual memory that needs to be mapped.
- Same as above but then we get BOOM and the following

that is not normal, it means the memory allocation failed.Make sure your system paging file is large enough (see [our wiki](https://github.com/MSOpenTech/redis/wiki/Memory-Configuration))

**Comment by ppanyukov** [comment]
@enricogior I don't think this is a page file size issue here.

When I say "Same as above but then we get BOOM" I mean that `VectoredHeapMapper` returns `EXCEPTION_CONTINUE_EXECUTION` (suggesting success) but then after that things would crash somewhere else (and I don't know exactly where they crash) -- see the image.

![image](https://cloud.githubusercontent.com/assets/300031/8881278/72e3a49c-3236-11e5-91e0-2eac294fd67c.png)

The crash would happen after we return from the marked line.

Besides, everything runs OK if I build for Debug. This does suggest some kind of pointer gets overwritten or something bug (more padding in Debug mode I assume), or am I completely wrong?

Memory-wise, I have the following:
- 32GB physical RAM
- Page file is set to be  48GB, although I'm not sure I've set all this in the right place, see the image

![image](https://cloud.githubusercontent.com/assets/300031/8881303/acd26364-3236-11e5-971d-4b57cc4c1887.png)

There is plenty of disk space everywhere too.

**Comment by enricogior** [comment]
@ppanyukov,
the system page file configuration looks good.

Do you have an existing AOF that can be used to reproduce the problem? If it's not sensitive data and you can provide a copy, it would help to find the root cause of the problem.
Thank you.

**Comment by ppanyukov** [comment]
@enricogior 

Sure, here are a couple: 
- 95MB: https://drive.google.com/file/d/0Bzo0KfFHv2cQcS15bHJMTi1xb1k/view?usp=sharing
- 190MB: https://drive.google.com/file/d/0Bzo0KfFHv2cQOVB5ZjVhWTJOc2M/view?usp=sharing

It's all random data there, nothing private.

If it helps, the keys are random GUID strings. The values are 64K binary arrays filled with random data using rand.

The command line I'm using to run redis is:

```
redis-server --loglevel verbose --maxheap 500000000 --maxmemory 134217728 --maxmemory-policy allkeys-lru --appendonly yes --hz 10 --save '""' --auto-aof-rewrite-percentage 100 --auto-aof-rewrite-min-size 64mb --no-appendfsync-on-rewrite yes
```

If you need a bigger AOF file, let me know.

**Comment by enricogior** [comment]
@ppanyukov 
thanks a lot for providing the AOF that reproduces the problem!
As you pointed out, it only happens with the Release build, while the Debug build is not effected so it took a little bit more effort to find the root cause: it's a tricky issue related to the way the memory is mapped in the child process, in some cases the virtual memory page that gets loaded is not visible to the fwrite() function that writes the data to disk and therefore it fails.

I have a fix that solves the problem but it needs to be polished and further tested before it can be released. Hopefully it will be ready in a day or two max.

Thanks again for you great contribution reporting the issue and providing detailed information and the sample to reproduce the bug!

**Comment by enricogior** [comment]
Hi @ppanyukov,
I've just pushed to the [2.8-staging branch](https://github.com/MSOpenTech/redis/commits/2.8-staging) the fix for this issue.
It will be included in the next official release that hasn't been scheduled yet (presumably early next week). In the meantime feel free to use the 2.8-staging code and let me know if the fix is working for you.
Thank you ;)

**Comment by ppanyukov** [comment]
@enricogior thanks for this, I will rerun this at some point.

Not sure if it's relevant:

Interestingly, looking at the patch and seeing it's apparently optimiser-related, we did try to eliminate any such optimisation bugs in our investigations, and we did compile for release with all optimisations off (both in compiler and linker).  And we still got the crashes every time. (obviously there are still subtle differences between debug and release configurations, but still quite curious).

**Comment by enricogior** [comment]
@ppanyukov 
the difference between the release and the debug was caused by this line:
    char x = _((char_)_Str + n);
that code was supposed to force a page fault so that the VEH would map the memory page.
That doesn't work with the release build because `x` is assigned but never used, therefore the compiler removes the entire line. Declaring `x` volatile will prevent the compiler from removing that line.
There was also another issue in the old code (hard-coded page size) but that was effecting both the release and the debug builds.

**Comment by ppanyukov** [comment]
@enricogior oh yes absolutely, I get that -- makes perfect sense and is pretty obvious once you see that line.  What I was referring to is that the `for` loop was optimised away in Release builds even when we switched off all compiler and linker optimizations we could think of.

But anyway, the fix looks good, I'm running the test off the [2.8-staging branch](https://github.com/MSOpenTech/redis/commits/2.8-staging) now and it's going well so far -- 2M sets done, 130GB of data, AOF rewrite is working as expected. No OOM crashes with `--maxheap` set to 10x `--maxmemory` either.

So I guess this can be closed now.

**Comment by enricogior** [comment]
@ppanyukov 
great, thank you ;)

**Comment by enricogior** [comment]
Hi @ppanyukov,
just to let you know that the fix is now part of the official [v2.8.2102 release](https://github.com/MSOpenTech/redis/releases/tag/win-2.8.2102).
Thank you.
